src/bin/pg_dump/dump_babel_utils.c

```diff
@@ -942,3 +944,179 @@ fixCopyCommand(Archive *fout, PQExpBuffer copyBuf, TableInfo *tbinfo, bool isFro
 	}
 	destroyPQExpBuffer(q);
 }
+
+/*
+ * hasSqlvariantColumn:
+ * Returns true if any of the columns in table is a sqlvariant data type column
+ */
+bool
+hasSqlvariantColumn(TableInfo *tbinfo)
+{
+	for (int i = 0; i < tbinfo->numatts; i++)
+		if (pg_strcasecmp(tbinfo->atttypnames[i],
+				quote_all_identifiers ? "\"sys\".\"sql_variant\"" : "sys.sql_variant") == 0)
+					return true;
+	return false;
+}
+
+/*
+ * getMbstrlen:
+ * returns the length of a multibyte string
+ */
+static int
+getMbstrlen(const char *mbstr, Archive *fout)
+{
+	int len = 0;
+	if (!mbstr)
+		return 0;
+	while (*mbstr){
+		mbstr += PQmblen(mbstr, fout->encoding);
+		len++;
+	}
+	return len;
+}
+
+/*
+ * fixCursorForBbfSqlvariantTableData:
+ * Prepare custom cursor for all Babelfish tables with atleast one sql_variant
+ * datatype column to correctly dump sql_variant data.
+ */
+int
+fixCursorForBbfSqlvariantTableData( Archive *fout,
+									TableInfo *tbinfo,
+									PQExpBuffer query,
+									int nfields,
+									int **sqlvar_metdata_pos)
+{
+	int orig_nfields = 0;
+	PQExpBuffer buf = createPQExpBuffer();
+
+	if (!isBabelfishDatabase(fout) || !hasSqlvariantColumn(tbinfo))
+		return 0;
+
+	*sqlvar_metdata_pos = (int *) pg_malloc0(3 * tbinfo->numatts * sizeof(int));
+	for (int i = 0; i < tbinfo->numatts; i++)
+	{
+		if (tbinfo->attisdropped[i])
+			continue;
+		if (tbinfo->attgenerated[i] && fout->dopt->column_inserts)
+			continue;
+
+		/* Skip TSQL ROWVERSION/TIMESTAMP column, it should be re-generated during restore. */
+		if (pg_strcasecmp(tbinfo->atttypnames[i],
+				quote_all_identifiers ? "\"sys\".\"rowversion\"" : "sys.rowversion") == 0 ||
+			pg_strcasecmp(tbinfo->atttypnames[i],
+				quote_all_identifiers ? "\"sys\".\"timestamp\"" : "sys.timestamp") == 0)
+			continue;
+
+		/*
+			* To find the basetype and bytelength of string data types we
+			* invoke sys.sql_variant_property and sys.datalength function on
+			* the sqlvariant column. These extra columns are added at the end of
+			* the select cursor query so that they do not interfere with
+			* expected dump behaviour.
+		*/
+		if (pg_strcasecmp(tbinfo->atttypnames[i],
+			quote_all_identifiers ? "\"sys\".\"sql_variant\"" : "sys.sql_variant") == 0)
+		{
+			appendPQExpBuffer(buf, ", sys.SQL_VARIANT_PROPERTY(%s, 'BaseType')", tbinfo->attnames[i]);
+			appendPQExpBuffer(buf, ", sys.datalength(%s)", fmtId(tbinfo->attnames[i]));
+			(*sqlvar_metdata_pos)[orig_nfields] = nfields;
+			nfields = nfields + 2;
+		}
+		orig_nfields++;
+	}
+	appendPQExpBuffer(query, buf->data);
```

Comment:
Wrong, should be : appendPQExpBuffer(query, "%s", buf->data).
Need to destroy buf as well: destroyPQExpBuffer(buf);

Reply:
Used appendPQExpBufferStr instead.